### PR TITLE
Reorganization of internal utilities and adding type hints for internal utils

### PR DESCRIPTION
**Pull request by freemansw1**
As part of my proposed efforts to add in type hints as a step toward xarray support (see my comment here: https://github.com/tobac-project/tobac/issues/143#issuecomment-1714202257), I've gone through and added type hints to the internal utility functions. 

I've also gone through and reorganized/cleaned up the internal utilities, breaking them into multiple different files, including specifically breaking off the Iris utilities into their own file for future deprecation/removal. 

Holding as a draft until 1.5.1 is released. 

* [x] Have you followed our guidelines in CONTRIBUTING.md? 
* [x] Have you self-reviewed your code and corrected any misspellings? 
* [x] Have you written documentation that is easy to understand?
* [x] Have you written descriptive commit messages? 
* [x] Have you added NumPy docstrings for newly added functions? 
* [x] Have you formatted your code using black? 
* [x] If you have introduced a new functionality, have you added adequate unit tests?
* [x] Have all tests passed in your local clone? 
* [ ] If you have introduced a new functionality, have you added an example notebook?
* [x] Have you kept your pull request small and limited so that it is easy to review? 
* [x] Have the newest changes from this branch been merged?

### CODE REVIEW

**Comment by w-k-jones** [comment]
Looks like the readthedocs build is failing due to using `collections.Callable` rather than `collections.abc.Callable`. Personally I've used `typing.Callable` before, but I have no preference for which is best

**Comment by freemansw1** [comment]
> Looks like the readthedocs build is failing due to using `collections.Callable` rather than `collections.abc.Callable`. Personally I've used `typing.Callable` before, but I have no preference for which is best

I was already using `typing.Callable` and not `collections.Callable`. I've switched to explicitly using `typing.Callable` instead of `from typing import Callable` to see if that fixes things. 

I'm not entirely sure why readthedocs required me to update the `.readthedocs.yml` in the first place, given that it otherwise works on our other PRs.

File: tobac/utils/internal/general_internal.py [hunk 1]
```diff
@@ -452,12 +467,15 @@ def njit_if_available(func, **kwargs):
         return func
 
 
-def find_vertical_axis_from_coord(variable_cube, vertical_coord=None):
+def find_vertical_axis_from_coord(
+    variable_cube: Union[iris.cube.Cube, xr.DataArray],
+    vertical_coord: Union[str, None] = None,
```
Comment (w-k-jones) [inline tobac/utils/internal/general_internal.py:472]: I find `|` clearer than `union` for type hints, but does that work for versions of python pre 3.10?

**Review by w-k-jones — APPROVED** [review]

File: tobac/utils/internal/general_internal.py (hunk 1 above)
Comment (freemansw1) [reply tobac/utils/internal/general_internal.py:472]: I looked into this and wasn't able to come to a firm conclusion as to when the `from __future__ import annotations` allow the `|` syntax. Unfortunately, I think matrix testing needs to move up the priority list to catch stuff like this.

Comment (freemansw1) [reply tobac/utils/internal/general_internal.py:472]: #353 addresses matrix testing

File: tobac/utils/internal/general_internal.py
[the remark this answers is not in the document]
Comment (JuliaKukulies) [reply tobac/utils/internal/general_internal.py:1]: Maybe `utils.internal.basic` or `utils.internal.basic` ?

**Review by JuliaKukulies — APPROVED** [review]

File: tobac/utils/internal/general_internal.py
[the remark this answers is not in the document]
Comment (freemansw1) [reply tobac/utils/internal/general_internal.py:1]: Agreed. I've renamed this now.

**Comment by freemansw1** [comment]
@JuliaKukulies @w-k-jones I've renamed `general_internal.py` to `basic.py`; let me know if that works for you, and we can get this merged.

**Comment by JuliaKukulies** [comment]
> @JuliaKukulies @w-k-jones I've renamed `general_internal.py` to `basic.py`; let me know if that works for you, and we can get this merged.

Perfect, happy for this to be merged!

**Pull request merged**